src/include/rho/jit/Optimization.hpp

```diff
@@ -0,0 +1,51 @@
+/*
+ *  R : A Computer Language for Statistical Data Analysis
+ *  Copyright (C) 2014 and onwards the Rho Project Authors.
+ *
+ *  Rho is not part of the R project, and bugs and other issues should
+ *  not be reported via r-bugs or other R project channels; instead refer
+ *  to the Rho website.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2.1 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  http://www.r-project.org/Licenses/
+ */
+
+#ifndef RHO_JIT_OPTIMIZATION_HPP
+#define RHO_JIT_OPTIMIZATION_HPP
+
+#include "rho/jit/llvm.hpp"
+
+namespace rho {
+namespace JIT {
+class BasicFunctionPass : public llvm::FunctionPass {
+ public:
+  BasicFunctionPass() : llvm::FunctionPass(pass_id) {}
+  BasicFunctionPass(BasicFunctionPass& other) = delete;
+  BasicFunctionPass& operator=(const BasicFunctionPass& other) = delete;
+  BasicFunctionPass(BasicFunctionPass&& other) = default;
+  BasicFunctionPass& operator=(BasicFunctionPass&& other) = default;
+
+  // Eliminates redundant calls to rho_runtime_setVisibility.  Any call that is
+  // postdominated by another is redundant.
+  // TODO(arunchauhan): Add more intra-procedural optimizations.
+  bool runOnFunction(llvm::Function& function);
```

Comment:
Since this class shouldn't be subclassed, I shouldn't have to create a virtual destructor.  Moreover, since the base class already defines its destructor virtual all the subclass destructors should automatically become virtual. I will add an explicit default one and also mark the method runOnFunction final.